libavcodec/vvc/dec.c

```diff
@@ -887,16 +999,192 @@ static int set_output_format(const VVCContext *s, const AVFrame *output)
+
+    return crc;
+}
+
+static int verify_crc(VVCContext *s, VVCFrameContext *fc)
+{
+    const VVCSPS  *sps   = fc->ps.sps;
+    const AVFrame *frame = fc->ref->frame;
+    uint8_t data[2] = { 0 };
+
+    int i, j, w, h;
+    uint32_t crc;
+
+    for (i = 0; frame->data[i]; i++) {
+        w = s->avctx->coded_width  >> sps->hshift[i];
+        h = s->avctx->coded_height >> sps->vshift[i];
+
+        crc = 0xFFFF;
+        for (j = 0; j < h; j++) {
+            const uint8_t *src = frame->data[i] + j * frame->linesize[i];
+            crc = cal_crc(crc, src, w << sps->pixel_shift);
+        }
+        crc = cal_crc(crc, data, FF_ARRAY_ELEMS(data));
+
+        av_log(s->avctx, crc == fc->sei.picture_hash.crc[i] ? AV_LOG_DEBUG : AV_LOG_ERROR ,
+            "Verifying CRC for frame with POC %d ", fc->ref->poc);
+        if (crc == fc->sei.picture_hash.crc[i]) {
+            av_log(s->avctx, AV_LOG_DEBUG, "plane %d - correct - %d\n",
+                i, crc);
+        } else {
+            av_log(s->avctx, AV_LOG_ERROR, "mismatching CRC of plane %d - %d != %d\n",
+                i, crc, fc->sei.picture_hash.crc[i]);
+            return AVERROR_INVALIDDATA;
+        }
+    }
+
+    return 0;
+}
+
+#define CAL_CHECKSUM(pixel) (checksum + ((pixel) ^ xor_mask)) & 0xFFFFFFFF
+static int verify_checksum(VVCContext *s, VVCFrameContext *fc)
+{
+    const VVCSPS  *sps   = fc->ps.sps;
+    const AVFrame *frame = fc->ref->frame;
+
+    int i, x, y, w, h;
+    int step = 1 << sps->pixel_shift;
+    uint32_t checksum;
+
+    for (i = 0; frame->data[i]; i++) {
+        w = s->avctx->coded_width  >> sps->hshift[i];
+        h = s->avctx->coded_height >> sps->vshift[i];
+
+        checksum = 0;
+        for (y = 0; y < h; y++) {
+            const uint8_t *src = frame->data[i] + y * frame->linesize[i];
+            for (x = 0; x < w; x++) {
+                int xor_mask = (x & 0xFF) ^ (y & 0xFF) ^ (x >> 8) ^ (y >> 8);
+                checksum = CAL_CHECKSUM(src[x << sps->pixel_shift]);
+                if (sps->pixel_shift)
+                    checksum = CAL_CHECKSUM(src[(x << sps->pixel_shift) + 1]);
+            }
+        }
+
+        av_log(s->avctx, checksum == fc->sei.picture_hash.checksum[i] ? AV_LOG_DEBUG : AV_LOG_ERROR ,
+            "Verifying CRC for frame with POC %d ", fc->ref->poc);
+        if (checksum == fc->sei.picture_hash.checksum[i]) {
+            av_log(s->avctx, AV_LOG_DEBUG, "plane %d - correct - %d\n",
+                i, checksum);
+        } else {
+            av_log(s->avctx, AV_LOG_ERROR, "mismatching CRC of plane %d - %d != %d\n",
+                i, checksum, fc->sei.picture_hash.checksum[i]);
+            return AVERROR_INVALIDDATA;
+        }
+    }
+
+    return 0;
+}
+
+static int frame_end(VVCContext *s, VVCFrameContext *fc)
+{
+    const AVFilmGrainParams *fgp;
+    int ret = 0;
+
+    if (fc->ref->needs_fg) {
+        av_assert0(fc->ref->frame_grain->buf[0]);
+        fgp = av_film_grain_params_select(fc->ref->frame);
```

Comment:
fgp can be NULL,

Reply:
Yeah. It is checked check_film_grain.
```
    if (fc->ref->needs_fg &&
        (fc->sei.common.film_grain_characteristics.present &&
            !ff_h274_film_grain_params_supported(fc->sei.common.film_grain_characteristics.model_id,
                fc->ref->frame->format) ||
            !av_film_grain_params_select(fc->ref->frame))) {
        av_log_once(s->avctx, AV_LOG_WARNING, AV_LOG_DEBUG, &s->film_grain_warning_shown,
            "Unsupported film grain parameters. Ignoring film grain.\n");
        fc->ref->needs_fg = 0;
    }
```

Reply:
Yes. But `needs_fg` is also set to be false. so it will not null here. The same as the `hevc`

Reply:
@QSXW 
Let us work together to make this upstream in the following weeks. 
As the first step, could you help rebase it to the latest?

Reply:
Yes sure Will update today ASAP.